# CV Problem

Our task is to build a image classifier that can classify the images<br>
in the test dataset to some classes of the images.<br>
We will train our model from training dataset given in the problem.<br>
Then fit the model into the testing data set and store it into a CSV file<br>
with the following details:<br>
    Test_image_index, predicted class.

Since the data given is in .pkl format we have to extract the data first so we import the Pickel library.<br><br>
<b>Intro to pickle format:</b><br>
Pickling is used to store python objects. This means things like lists, dictionaries, class objects, and more.<br>
while saving the data in PKL format it's nature is also saved.For ex. if we save a dictionary to a pickle format file<br>
then when loading it into a variable it's nature will be dictionary no need to worry about that.<br>
In simple it is saving the pyhton objects in stream of bytes which can then be accessed backed easily.<br>

In [4]:
import pickle

First we have to deserialise the training data into a python object.And then use it to our testing data.

The motto behind the pickling is suppose if we are having a very large dataset in any format say mysql or .csv file<br>
When extracting that data to use it in our program will take a lot of time instead pickling will not.<br>
You can refer further to pickling to the following link:<br>
    <a href="https://pythonprogramming.net/python-pickle-module-save-objects-serialization/">Know more about Pickling</a>

Now we all know about some basic library
that is <b>numpy</b> or numerical python library which consisting of multidimensional array<br>
objects and a collection of routines for processing those arrays.

In [25]:
import numpy as np
import csv #for storing the predicted result into the CSV file

<b>Now we will use TensorFlow library which is specially used for branch of machine learning<br>
that is Deep Learning.</b>

In [6]:
import tensorflow as tf
from tensorflow import keras

Want to know more about tensorflow here is the link:<br>
    <a href="https://cv-tricks.com/artificial-intelligence/deep-learning/deep-learning-frameworks/tensorflow/tensorflow-tutorial/">Know more Tensorflow</a>

Now we get the train dataset from the train_label.pkl file given in the problem.

In [10]:
with open('train_label.pkl','rb') as f:
    #As our file is stored in streams of bytes.We open it as read in byte mode.
    train_label=pickle.load(f)

Next we extract the train Image.

In [11]:
with open('train_image.pkl','rb') as f:
    train_image=pickle.load(f)

Now we put the train image to the numpy array.So that we can use tensorflow to train our model.

In [12]:
train_image=np.array(train_image)

Let's visualise it more better.

In [13]:
#A method in matplot library to know the dataset structure.
train_image.shape

(8000, 784)

In [37]:
len(train_label)

8000

Now look to the above results.<br>
The train label lenght is 8000 meaning there are 8000 images.<br>
784 is length of the train image which is basically a grayscale 28*28 pixels image.<br>
What we are going to do is to store this image into a 3d array of 8000*28*28.

In [18]:
train_image=train_image.reshape(8000,28,28)

Intentions behind doing this:
<br>
We all know about the neural network.Which takes some inputs,process it according<br>
to a certain algorithm and gives the output.We are passing a image into the inputs which<br>
are the nodes of total no 28*28=784.Training our model according to it and then predicting<br>
the resulting output.<br>
If you want to know more about this then go to the following link.<br>
<a href="http://neuralnetworksanddeeplearning.com/chap1.html">Neural networks</a><br>
<img src="2.jpg">


Now we are going to reshpae our test data as of the train data.

In [15]:
with open('test_image .pkl','rb') as f:
    test_image=pickle.load(f)
test_image=np.array(train_image)
test_image=test_image.reshape(8000,28,28)
test_image.shape

(8000, 28, 28)

Till now we were shaping our data.<br>
Algorithm our network is using:<br>
    <ol>
    <li>We are passing the image as pixels of 28*28 i.e  784 into the inputs of neural network<br>
    in form of bits either 0 or 1.So we have divide each pixel of our image by 255.</li>
    <li>As explained below then we going to get which class a particular image belongs to</li>
    <li>Basically we are first training our model with the training dataset by passing the images<br>
    of training dataset and then as given classes.</li>
    <li>Next fit out model to test data and save it in a .csv(comma seperated value file)</li>
    </ol>

In [19]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation=tf.nn.relu),
    keras.layers.Dense(10, activation=tf.nn.softmax)
])
#Passing our data
model.compile(optimizer='adam', 
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
#fitting our model
model.fit(train_image, train_label, epochs=50)

Epoch 1/50
8000/8000 [==============================] - 2s 209us/sample - loss: 12.0986 - acc: 0.2494
Epoch 2/50
8000/8000 [==============================] - 1s 122us/sample - loss: 12.0886 - acc: 0.2500
Epoch 3/50
8000/8000 [==============================] - 1s 121us/sample - loss: 12.0886 - acc: 0.2500
Epoch 4/50
8000/8000 [==============================] - 1s 123us/sample - loss: 12.0886 - acc: 0.2500
Epoch 5/50
8000/8000 [==============================] - 1s 147us/sample - loss: 12.0886 - acc: 0.2500
Epoch 6/50
8000/8000 [==============================] - 1s 119us/sample - loss: 12.0886 - acc: 0.2500
Epoch 7/50
8000/8000 [==============================] - 1s 119us/sample - loss: 12.0886 - acc: 0.2500
Epoch 8/50
8000/8000 [==============================] - 1s 121us/sample - loss: 12.0886 - acc: 0.2500
Epoch 9/50
8000/8000 [==============================] - 1s 134us/sample - loss: 12.0886 - acc: 0.2500
Epoch 10/50
8000/8000 [==============================] - 1s 140us/sample - loss: 1

Now as our model is trained by passing the train dataset into the model.We now predicts our test data<br>
and create a 2d matrix to store it.

In [24]:
#predicting our test data
predict_test_data=model.predict(test_image)
image_index=list(range(0,8000))
#index of image is a list
predicted_class=[]
#for storing the predicted values.
for i in range (8000):
    predicted_class.append(np.argmax(predict_test_data[i]))
#Storing it to a CSV file
d_matrix=list(zip(image_index,predicted_class))
#inserting the values
d_matrix.insert(0,('Test_image_index','predicted_class'))

Now storing to csv file.

In [27]:
with open('RamdevGodara.csv', 'w') as f:
    writer = csv.writer(f)
    writer.writerows(d_matrix)

<b><h3>Conclusion:</h3><br>
    Just build a image classifier using a very simple neural network whose methods are even defined in Tensorflow library.</b>